docs/engineering-principles.md

```diff
@@ -0,0 +1,48 @@
+1. Security
+2. Stability
+3. Safety
+4. Privacy
+5. Maintainability
+6. Usability
+
+
+## Principles
+### The Principle of Least Authority
+
+Provide application code the absolute minimal amount of authority required to perform its essential functions. Wherever appropriate, break down large methods with broad authority into smaller functions with minimal authority. Minimizing what a method *can* do reduces the scope of what could possibly go wrong. When something does go wrong, small isolated methods, written with the least amount of authority required, are easier to troubleshoot.
+
+### The Principle of Core Purity
+
+Whenever possible prefer small, pure functions. Pure functions are those that given the same inputs will always return the same output, and that have no side effects. Where side effects must occur, keep the core of the application code pure and move imperative methods to the outer edges of the application layer.
+
+
+### The Principle of Incremental Change
+
+To minimize the risk to the stability and security of the application, changes to the code should be incremental and as small as possible. Large changes to the codebase become difficult to review effectively and make the commit history less itemized. Furthermore, incremental change promotes better test coverage and makes it easier to require tests to cover the introduced changes.
+
+### The Principle of Incremental Decentralization
+
+We may sometimes ship with centralized services by default, but if we must, users should be able to opt out of any of these during onboarding, ideally easily swapping them out for decentralized alternatives. When we build features that rely upon centralized services we must build them in such a way where the experience isn't completely broken when the service is disrupted. We are trying to build a decentralized future and we must make choices that move us directionally towards those goals.
+
+### The Principle of Self-Soverign Identity
+
+We believe in the right of our users to own their own identity and data. We must go to great lengths to protect that right. Where we have features or services that might infringe upon that right we must treat those features as opt in, not opt out. When we do build features that make user data known to us we must not collect any personally identifiable information, and any data that can be tied to any chain must be collected in an anonymity protecting way. This principle must be extended to cover all third party code as well.
+
+### The Principle of Auditability
+
+All code contributions must be reviewed thoroughly for code quality, usability and adherence to these principles. Code contributions must be kept as small as possible (see Principle of Incremental Change) to promote auditability. Avoid long running feature branches because they are not auditable, and pose a risk to the stability and security of the product.
```

Comment:
Is this saying that we, as a team, should commit to thoroughly auditing our work? Or that we should be committed as a team to making our work easy to audit? Or both?

Reply:
It feels like this principle and the "incremental change" are touching upon some shared idea. I know they're different but they are closely related :thinking:

Reply:
Both, and while I agree that incremental change is a requirement for auditability, they have different intentions. One is focusing on incremental change as a safety mechanism and to prevent drastic changes in behavior in a single pull request. The other is about making sure we both review code thoroughly and that we make it easy for our code to be reviewed. I'm going to update the definitions of these slightly to further the differences between them, rather than condensing into one.